libsql-server/src/pager.rs

```diff
@@ -0,0 +1,319 @@
+    let allocator = PAGER_CACHE.get_or_init(|| Allocator::new(page_size as usize, extra_size as usize, PAGER_CACHE_SIZE.load(SeqCst)));
+    let pager = Pager {
+        alloc: allocator,
+        pages: HashMap::new().into(),
+    };
+
+    Box::leak(Box::new(pager)) as *mut Pager as *mut _
+}
+
+extern "C" fn cache_size(_cache: *mut sqlite3_pcache, _size: c_int) { }
+extern "C" fn page_count(cache: *mut sqlite3_pcache) -> c_int {
+    let cache = unsafe { &*(cache as *mut Pager) };
+    cache.pages.len() as _
+}
+
+extern "C" fn fetch(cache: *mut sqlite3_pcache, key: u32, create_flag: c_int) -> *mut sqlite3_pcache_page {
+    let cache = unsafe { &mut *(cache as *mut Pager) };
+    let pages = &mut cache.pages;
+    match pages.get_mut(&key) {
+        Some(page) => {
+            page.pin(key as u16);
+            (*page) as *mut _ as *mut _
+        }
+        None => {
+            // try to find an unpinned page
+            match pages.extract_if(|_, p| !p.is_pinned()).next() {
+                Some((_, page)) => {
+                    page.pin(key as u16);
+                    let ptr = page as *mut _; 
+                    pages.insert(key, page);
+                    ptr as *mut _
+                },
+                None if create_flag == 0 => std::ptr::null_mut(),
+                None if create_flag != 0 => {
+                    // try alloc one from global pool
+                    match cache.alloc.alloc() {
+                        Some(page) => {
+                            page.pin(key as u16);
+                            let ptr = page as *mut _;
+                            assert!(pages.insert(key, page).is_none());
+                            ptr as *mut _
+                        },
+                        None => {
+                            std::ptr::null_mut()
+                        },
+                    }
+                }
+                None => unreachable!(),
+            }
+        },
+    }
+}
+
+extern "C" fn unpin(cache: *mut sqlite3_pcache, page: *mut sqlite3_pcache_page, discard: c_int) {
+    let page: &mut Page = unsafe { &mut *(page as *mut Page) };
+    let cache = unsafe { &mut *(cache as *mut Pager) };
+    let pages = &mut cache.pages;
+
+    let key = page.unpin() as u32;
+
+    if discard != 0 {
+        let page = pages.remove(&key).expect("missing page");
+        cache.alloc.free(page);
+    }
+}
+
+extern "C" fn rekey(cache: *mut sqlite3_pcache, data: *mut sqlite3_pcache_page, old_key: u32, new_key: u32) {
+    let cache = unsafe { &mut *(cache as *mut Pager) };
+    let _new_page = unsafe { &mut *(data as *mut Page) };
+    let pages = &mut cache.pages;
+    let page = pages.remove(&old_key).expect("missing page when rekeying");
+    page.pin(new_key as u16);
+    if let Some(page) = pages.insert(new_key, page) {
+        assert!(!page.is_pinned());
+        cache.alloc.free(page);
+    }
+}
```

Comment:
@psarna It's not clear to me what I'm supported to do with new page. https://sqlite.org/c3ref/pcache_methods2.html suggests that we are just moving a page, but then why is that necessary to pass the data as well?

Reply:
my undertanding is that data is a ref to the old key's value already

Reply:
let me browse the C code and see

Reply:
@MarinPostma ok, so the stock implementation is here: https://github.com/tursodatabase/libsql/blob/ac6bf3a4a9eca21e004654b82a04149b1835850b/libsql-sqlite3/src/pcache1.c#L1113-L1147

... and I don't follow the question about why we need to pass the page data -- we need its metadata, which in the stock implementation is really a pointer to https://github.com/tursodatabase/libsql/blob/ac6bf3a4a9eca21e004654b82a04149b1835850b/libsql-sqlite3/src/pcache1.c#L117-L127 . `PgHdr1` has an intrusive hashtable built-in, so when we update the key and put it in another place in the data structure, we need the page instance to update the internal pointers. So I guess it's just an implementation detail, and since your hashmap isn't intrusive, you just don't need to touch the page pointer either.